internal/pipe/pipe.go

```diff
@@ -0,0 +1,131 @@
+package pipe
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/newrelic/newrelic-cli/internal/utils"
+)
+
+type pipeReader interface {
+	ReadPipe() (string, error)
+}
+
+type stdinPipeReader struct {
+	input io.Reader
+}
+
+func (spr stdinPipeReader) ReadPipe() (string, error) {
+	text := ""
+	var err error = nil
+	scanner := bufio.NewScanner(spr.input)
+
+	for scanner.Scan() {
+		text = fmt.Sprintf("%s%s ", text, strings.TrimSpace(scanner.Text()))
+	}
+
+	if scanErr := scanner.Err(); scanErr != nil {
+		err = scanErr
+	}
+
+	return strings.TrimSpace(text), err
+}
+
+func jsonToFilteredMap(r string, selectors []string) ([]map[string]string, error) {
+	text := r
+
+	if !gjson.Valid(text) {
+		return nil, errors.New("invalid JSON received by stdin")
+	}
+
+	// always start with an array of values
+	if strings.HasPrefix(text, "{") {
+		text = fmt.Sprintf("[ %s ]", text)
+	}
+
+	// returns []gjson.Result
+	jsonArray := gjson.Parse(text).Array()
+
+	resultsArray := make([]map[string]string, len(jsonArray))
+
+	for index, resultObj := range jsonArray {
+		resultMap := make(map[string]string)
+		for _, selector := range selectors {
+			if value := resultObj.Get(selector); value.Exists() {
+				// Convert every value to a string
+				resultMap[selector] = value.String()
+			}
+		}
+		resultsArray[index] = resultMap
+	}
+
+	return resultsArray, nil
+}
+
+func readStdin(pipe pipeReader, selectorList []string) ([]map[string]string, error) {
+	jsonString, pipeErr := pipe.ReadPipe()
+	if pipeErr != nil {
+		return nil, pipeErr
+	}
+
+	filteredMap, mapErr := jsonToFilteredMap(jsonString, selectorList)
+	if mapErr != nil {
+		return nil, mapErr
+	}
+
+	return filteredMap, nil
+}
+
+func pipeInputExists() bool {
+	fi, err := os.Stdin.Stat()
+	if err != nil {
+		return false
+	}
+	return (fi.Mode() & os.ModeCharDevice) == 0
+}
+
+func getPipeInputInnerFunc(pipe pipeReader, pipeInputExists bool, acceptedPipeInput []string) map[string][]string {
+	if pipeInputExists {
+		pipeInput := map[string][]string{}
+		inputArray, err := readStdin(pipe, acceptedPipeInput)
+		if err != nil {
+			utils.LogIfError(err)
+			return map[string][]string{}
+		}
+		for _, key := range acceptedPipeInput {
+			var collectedItemsForKey []string
+			for _, value := range inputArray {
+				collectedItemsForKey = append(collectedItemsForKey, value[key])
+			}
+			pipeInput[key] = collectedItemsForKey
+		}
+		return pipeInput
+	}
+	return map[string][]string{}
+}
+
+func getPipeInputFactory(pipe pipeReader, predicate func() bool) func([]string) {
+	return func(acceptedPipeInput []string) {
+		pipeInput = getPipeInputInnerFunc(pipe, predicate(), acceptedPipeInput)
+	}
+}
+
+var GetInput = getPipeInputFactory(stdinPipeReader{input: os.Stdin}, pipeInputExists)
+
+var pipeInput map[string][]string = map[string][]string{}
+
+func Get(inputKey string) ([]string, bool) {
```

Comment:
This method looks to be called once unless I overlooked another place.  What do you think about folding this logic into the `Exists()` method?

Reply:
The Get method would be called everytime a command needs to get a value. It's called once because I've only used this module on one command. It's just good, clean, self-explanatory API design.